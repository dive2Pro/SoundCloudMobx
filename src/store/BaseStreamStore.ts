import performanceStore from './PerformanceStore';
import {
  action, observable, runInAction
  , computed, ObservableMap, autorun
  , IReactionDisposer, IObservableArray
} from 'mobx';

import {
  ITrack
} from '../interfaces/interface';
import { GENRES } from '../constants/trackTypes'

export interface IBaseActStore {
  isLoading: boolean;
  currentGenre: string
  isError: (genre: string) => boolean
}
abstract class BaseStreamStore<T> implements IBaseActStore {
  static displayName = 'BaseAct'
  // 记得初始化
  protected itemsMap = new ObservableMap<T[]>();
  @observable protected nextHrefsByGenre = new ObservableMap<string>();
  protected isLoadingByGenre = {
    get: performanceStore.getLoadingStateWidthKey
  }
  protected isErrorsMap = new ObservableMap<boolean>()

  // 排序和搜索过滤
  @observable filterType: string
  @observable filterTitle: string
  @observable sortType: string
  @observable filteredTracks: ITrack[] = []
  @observable currentGenre: string = GENRES[0]

  autorunHandle: IReactionDisposer;

  @computed get isLoading(): boolean {
    return this.isLoadingByGenre.get(this.currentGenre)
  }
  
  @computed get nextHref() {
    return this.nextHrefsByGenre.get(this.currentGenre) || ''
  }

  @action setFilterTitle(title: string) {
    this.filterTitle = title;
  }

  @action setSortType(type: string) {
    this.sortType = type;
  }

  @action setFilterType(filterType: string = '') {
    this.filterType = filterType;
  }

  setLoadingByGenre(genre: string, loading: boolean) {
    // this.isLoadingByGenre.set(genre, loading);
    performanceStore.setLoadingStateWithKey(genre, loading);
  }

  @action setNextHrefByGenre(genre: string, nextHref: string) {
    this.nextHrefsByGenre.set(genre, nextHref)
  }

  @action setGenre(genre: string) {
    // genre = genre.toLocaleLowerCase();
    this.currentGenre = genre;
    performanceStore.setCurrentGenre(this.currentGenre)
    this.initFilterFunction(genre);
  }

  @computed get currentItems(): T[] {
    const hasItems = this.itemsMap.has(this.currentGenre)
    if (!hasItems) {
      this.itemsMap.set(this.currentGenre,[])
    }
    const item = this.itemsMap.get(this.currentGenre);
    return <T[]>item
  }

  isError = (genre: string): boolean => {
    return performanceStore.isError(genre)
  }

  @action protected catchErr = (err: any, genre: string) => {
    performanceStore.catchErr(err, genre)
    console.error(err, genre)
  }
  protected resetErrorWithType = (type: string) => {
    performanceStore.resetErrorsMap(type)
  }
  protected filterByFilterType(fs: T[]): T[] {
    return fs;
  }
  abstract transToTracks(act: T[]): ITrack[];

  private async filterFunc(activities: T[], filterTitle: string, sortType: string, filterType: string) {
    let fs: ITrack[] = []
    const filterByTypes = await this.filterByFilterType(activities)
    // console.log(filterByTypes)
    fs = await this.transToTracks(filterByTypes)
    fs = await this.filterByFilterTitle(fs);
    fs = await this.filterBySortType(fs);

    runInAction('set-filteredTracks', () => {
      (<IObservableArray<ITrack>>this.filteredTracks).replace(fs)
    })
  }
  private filterBySortType(fs: ITrack[]) {
    let temp = fs.slice();
    if (!!this.sortType) {
      temp = fs.sort((p, n) => {
        let pCount = p[this.sortType];
        let nCount = n[this.sortType];
        pCount = !Number.isNaN(pCount) ? pCount : 0;
        nCount = !Number.isNaN(nCount) ? nCount : 0;
        return nCount - pCount;
      })
    }
    return temp
  }
  private filterByFilterTitle(fs: ITrack[]) {
    let temp = fs.slice();
    if (!!this.filterTitle) {
      temp = fs.filter(item => {
        return item.title.indexOf(this.filterTitle) > -1
      })
    }
    return temp
  }


  private initFilterFunction(type: string) {

    if (this.autorunHandle) {
      this.autorunHandle();
    }
    let has = this.itemsMap.has(type);
    if (!has) {
      this.itemsMap.set(type, []);
    }
    this.autorunHandle = autorun(() => {
      this.filterFunc(<T[]>this.itemsMap.get(type), this.filterTitle, this.sortType, this.filterType)
    })
  } 
}
export default BaseStreamStore