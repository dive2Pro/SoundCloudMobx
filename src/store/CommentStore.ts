import {
  ObservableMap, observable, computed, action, runInAction
  // , createTransformer
} from 'mobx'
import { apiUrl, unauthApiUrl } from '../services/soundcloundApi'
import { ITrack } from '../interfaces/interface';
import  performanceStore  from './PerformanceStore'
import  sessionStore  from './SessionStore'
import { FETCH_COMMENTS } from '../constants/fetchTypes';
import { POST_TARCK_COMMENT } from '../constants/requestTypes';

import { RaceFetch as fetch } from '../services/Fetch'

export interface IComment {
  id: number,
  created_at: string,
  user_id: number,
  track_id: number,
  timestamp: number,
  body: string
  uri: string,
  user: {
    id: number,
    permalink: string,
    username: string,
    uri: string,
    permalink_url: string,
    avatar_url: string
  }
}



export class CommentStore {
  @observable currentTrack: ITrack|null = null
  @observable private nextHrefsByTrack
   = new ObservableMap<string>();
  
  private commentsByTracks = new ObservableMap<IComment[]>();

  constructor() {
    performanceStore.setLoadingStateWithKey(FETCH_COMMENTS, false)
    performanceStore.setLoadingStateWithKey(POST_TARCK_COMMENT, false)
  }
 

  


  @computed get isLoadingMoreComment() { 
    return performanceStore.getLoadingStateWidthKey(FETCH_COMMENTS);
  }

  @computed get commentsCount() {
    return this.currentTrackComments.length
  }

  @computed get currentTrackComments() {
    if (!this.currentTrack) { return []; }
    const id =this.currentTrack.id + ''
    const has = this.commentsByTracks.has(id)
    if (!has) {
      this.commentsByTracks.set(id,[])
    }
    return this.commentsByTracks.get(id)||[]
  }
  @computed get currentCommentNextHref(): string {
    return this.currentTrack &&this.nextHrefsByTrack.get(this.currentTrack.id + '') || ''
  }

  setLoadingState(type: string, loading: boolean) {
    performanceStore.setLoadingStateWithKey(type, loading);
  }

  @action setCurrentTrack(track: ITrack) {
    this.currentTrack = (track);
    
  }

  @action async fetchMoreComments(nextHref?: string) {
    if (this.isLoadingMoreComment) { return }
    nextHref = this.currentCommentNextHref
    if (nextHref == 'EMPTY'||!this.currentTrack) {
      return
    }
    const { id } = this.currentTrack
    const keyId = id + ''
    let url = nextHref
      ? nextHref : apiUrl(`tracks/${id}/comments?linked_partitioning=1&limit=50&offset=0`, '&');

    this.setLoadingState(FETCH_COMMENTS, true)

    try {
       const data: any = await fetch(url); 
       const oldData = this.commentsByTracks.get(keyId)
       runInAction(() => {
        if (oldData) {
          oldData.push(...data.collection);
        } else {
          this.commentsByTracks.set(keyId, data.collection);
        }
        this.nextHrefsByTrack.set(keyId, data.next_href || 'EMPTY');
      })
    } catch (err) {
      performanceStore.catchErr(err, FETCH_COMMENTS)
    } finally {
      this.setLoadingState(FETCH_COMMENTS, false)
    }
  }

  // todo 用decortator来处理 try_catch  
  async submitReplay(id: number, msg: string) {
    const user = sessionStore.user
    if (!user) {
      return
    } 
    
    const url = unauthApiUrl(`/tracks/${id}/comments`, '&')
    try {
      this.setLoadingState(POST_TARCK_COMMENT, true)
      // const data = await fetch(url, {
      // method: 'POST', 'Content-Type': `application/x-www-form-urlencoded; charset=UTF-8 `
      // })      
      const data = await new Promise((resolve, reject) => {
        setTimeout(() => {
          resolve(rawData(msg))
         
        }, 500)
      }) 
      this.pushLastReplay(data)
      
    } catch (err) {
      performanceStore.catchErr(err, POST_TARCK_COMMENT)
    } finally {
      this.setLoadingState(POST_TARCK_COMMENT, false)
    }

  }
  @action private pushLastReplay(data) {
    if (!this.currentTrack) { 
      return
    }
    const comments = this.currentTrackComments;  
    comments.unshift(data)
  }
}
  export const rawData= (msg:string):IComment => ({
            kind: 'comment'
            , id: 341101782,
            created_at: '2017/04/13 16:23:38 +0000',
            user_id: 278227204,
            track_id: 280186218,
            timestamp: 450,
            body: `@huang-alex-635833920: ${msg}`,
            uri: 'https://api.soundcloud.com/comments/341101782',
            user: {
              id: 278227204
              , kind: 'user',
              permalink: 'huang-alex-635833920',
              username: 'Huang Alex',
              last_modified: '2017/03/31 08:35:22 +0000',
              uri: 'https://api.soundcloud.com/users/278227204',
              permalink_url: 'http://soundcloud.com/huang-alex-635833920',
              avatar_url: 'https://i1.sndcdn.com/avatars-000286355267-ibfmck-large.jpg'
            }
  })
          
export default new CommentStore();