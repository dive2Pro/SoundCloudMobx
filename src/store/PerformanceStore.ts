import {
  observable
  , action
  , IObservableArray
  , ObservableMap
  , computed
  , expr
} from 'mobx'

export class PerformanceStore {

   __breaks={
    $breakMedium:700
  }

  @observable windowWidth = window.innerWidth;

  onceLoadingIsAllSettle: boolean;
  scrollLimitByGenre = new ObservableMap<number[]>()
  @observable genre: string
  // 这个用来记录当前应该被 player 进行毛玻璃效果处理的dom id
  @observable glassNode: string
  @observable scrollY: number
  isLoadingsByKey = new ObservableMap<boolean>();
  isErrorsMap = new ObservableMap<boolean>()
  
  @computed get scrollLimit(): number[] {
    return this.scrollLimitByGenre.get(this.genre) || [];
  }

  @action setCurrentGenre(genre: string) {
    this.genre = genre;
    if (!this.scrollLimitByGenre.get(genre)) {
      this.scrollLimitByGenre.set(
        genre,
        [window.innerHeight, window.innerHeight]
      )
    }
  }

  getLoadingState(type: string): boolean {
    return this.isLoadingsByKey.get(type) || false
  }

  @action setScrollLimit(...limit: number[]) {
    const map = this.scrollLimitByGenre.get(this.genre);
    if (map) {
      (<IObservableArray<number>>map).replace(limit);
    } else {
      this.scrollLimitByGenre.set(this.genre, limit)
    }
  }

  @action setCurrentGlassNodeId(id: string) {
    this.glassNode = id;
  }

  @action setScrollY(y: number) {
    this.scrollY = y;
  }

  @action setLoadingStateWithKey = (key: string, loading: boolean) => {
    this.isLoadingsByKey.set(key, loading)
  }

  getLoadingStateWidthKey = (key: string) => {
    return this.isLoadingsByKey.get(key) || false
  }

  /**
   * 当前所以请求完毕
   */
  @computed get allLoadingIsSettle(): boolean {
    const allSettle = expr(() =>
      Array.from(this.isLoadingsByKey.values()).every(v => v === false))
    return allSettle
  }


  isError = (genre: string): boolean => {
    return this.isErrorsMap.get(genre) || false
  }

  @action catchErr = (err: any, genre: string) => {
    this.isErrorsMap.set(genre, true);
  }

  @action resetErrorsMap = (fetchType: string) => {
    this.isErrorsMap.set(fetchType, false);
  }

  @action setWindowSize(size){
    this.windowWidth = size
  }

  @computed get isUnderLarge(){
    return this.windowWidth < 1200
  }

  @computed get iUnderMedium(){
    return this.windowWidth < 680
  }

  @computed get iUnderHandsets(){
    return this.windowWidth < 475
  }

}

export default new PerformanceStore()