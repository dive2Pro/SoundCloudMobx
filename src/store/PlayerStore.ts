import { observable, action, computed } from 'mobx';
import { addClientId } from '../services/soundcloundApi'
import { ITrack } from '../interfaces/interface'

const isEqual =require('lodash/isEqual')
export class PlayerStore {
  @observable playingTrack: ITrack
  @observable isPlaying: boolean = false;
  @observable playList: ITrack[] = [];

  @observable isPlaylistOpen: boolean = false;
  @observable isShuffleMode: boolean = false;
  @observable isVolumeOpen: boolean = false

  @observable volume: number = 0.25

  @action setVolume(v: number) {
    this.volume = v;
  }

  @action setPlayingTrack(track: ITrack | number) {
    if (typeof track === 'number') {
      track = this.playList[track];
    }

    if (isEqual(this.playingTrack, track) && this.isPlaying) {
      this.isPlaying = false;
      
    } else {
      this.playingTrack = track;
      this.isPlaying = true;
      this.addToPlaylist(track);
    }
  }

  @action togglePlaying() {
    if(!this.playingTrack){
        this.playNextTrack(1)
    }else{
      this.isPlaying = !this.isPlaying;

    }

  }
  @action toggleVolumeOpen(open?: boolean) {
    if (open != null) {
      this.isVolumeOpen = open
    } else {
      this.isVolumeOpen = !this.isVolumeOpen;
    }
  }

  @action toggleShuffleMode() {
    this.isShuffleMode = !this.isShuffleMode
  }


  @action clearPlaylist() {
    this.playList = [];
  }
  @action removeFromPlaylist(...tracks: ITrack[]) {
    tracks.forEach(t => {
      const index = this.playList.indexOf(t)
      this.playList.splice(index, 1)
    })
  }

  @action togglePlaylistOpen(open?: boolean) {
    if (open != null) {
      this.isPlaylistOpen = open
    } else {
      this.isPlaylistOpen = !this.isPlaylistOpen;
    }
  }

  @computed get getPlaylistCount() {
    return this.playList && this.playList.length;
  }
  @computed get playingUrl() {
    if (this.playingTrack) {
      let url = this.playingTrack.uri + '/stream';
      url = addClientId(url, '?');
      return url
    }
    return ''
  }
  @action playNextTrack(diff: number): boolean {
    if (this.getPlaylistCount < 1) {
      return false;
    }
    
    let nextTrack: ITrack | undefined;
    if (this.isShuffleMode) {
      const newList = this.playList.slice();
      const deleIndex = newList.indexOf(this.playingTrack)
      newList.splice(deleIndex, 1);
      // tslint:disable-next-line:radix
      nextTrack = newList[parseInt(Math.random() * 10 + '') % newList.length];
    } else {
      let changedIndex = 0;
      if (this.playingTrack) {
        changedIndex = this.playList.indexOf(this.playingTrack) + diff;
      }

      nextTrack = this.playList[changedIndex];
    }
    if (nextTrack != null) {
      this.setPlayingTrack(nextTrack);
      return true
    }

    return false
  }

  addToPlaylist(tracks: ITrack | ITrack[]) {
    let ts:any = tracks
    if (ts.slice&&Array.isArray(ts.slice())) {
      (<ITrack[]>tracks).slice().forEach((t, i) => {
        i === 0 && this.setPlayingTrack(t)
        this.pushToPlayerlist(t)
      })
    } else {
      this.pushToPlayerlist(<ITrack>tracks)
    }
  }

  @action private pushToPlayerlist(t: ITrack) {
    if (!this.playList.some((item)=>item.id===t.id)) {
      this.playList.push(t)
    }
  }
}

export default new PlayerStore();