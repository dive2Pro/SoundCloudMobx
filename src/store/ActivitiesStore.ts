import { action, computed, runInAction } from 'mobx'
import  BaseStreamStore  from './BaseStreamStore';
import { IActivitiesItem, ITrack } from '../interfaces/interface';
import { FETCH_ACTIVITIES } from '../constants/fetchTypes';
import  performanceStore from './PerformanceStore';
import { addAccessToken, apiUrl } from '../services/soundcloundApi';
import { RaceFetch as fetch } from '../services/Fetch'

export class ActivitiesStore extends BaseStreamStore<IActivitiesItem>  {

  constructor() {
    super()
    this.setGenre(FETCH_ACTIVITIES)
  }

  @action setNextActivitiesHref(nextHref: string) {
    this.setNextHrefByGenre(FETCH_ACTIVITIES, nextHref)
  }

  @action addActivities(arr: IActivitiesItem[]) {
    const itemsLength = this.currentItems.length;
    this.currentItems.splice(itemsLength, 0, ...arr);
    // const t : any= []
    // const iis:any = t.splice(0, 0, ...arr)
    // console.log(
    //   // t, '-------------',
    //   this.currentItems)
  }

  transToTracks(items: IActivitiesItem[]): ITrack[] {
    return items.map(this.getAllTrackFromActivity)
      .filter(item => item != null);
  }

  getAllTrackFromActivity(act: IActivitiesItem) {
    return act.origin
  }

  @computed get tracks() {
    return this.currentItems && this.currentItems.map(this.getAllTrackFromActivity)
  }

  filterByFilterType(fs: IActivitiesItem[]) {
    let temp = fs.slice();
    if (!!this.filterType) {
      temp = fs.filter(item => {
        return item.type === this.filterType
      })
    }
    return temp;
  }
  async filterActivities(arr: IActivitiesItem[]) {
    const filterArr = await arr.filter(item => {
    const b = this.currentItems.some(active =>
        active.created_at === item.created_at)
    return !b;
    })
    this.addActivities(filterArr);
  }

  setLoadingActivities(b: boolean) {
    this.setLoadingByGenre(FETCH_ACTIVITIES, b);
  }

  async fetchNextActivities(first?: boolean) {
     
    if (this.isLoading) {
       
      return  
    }
    
    if (first && this.currentItems.length > 0) {
      return
    }

    performanceStore.setCurrentGenre(this.currentGenre)
    await this.fetchActivities(this.nextHref);
  }
  @action private async  fetchActivities(nextHref?: string) {
    if (this.isLoading) { return }
    let activitiesUrl = nextHref ? addAccessToken(nextHref, '&') :
      apiUrl(`me/activities?limit=50`, '&')
    try {
      this.setLoadingActivities(true);
      const data: any = await fetch(activitiesUrl)
      runInAction(() => {
        this.setNextActivitiesHref(data.next_href)
        this.filterActivities(data.collection);
      })
    } catch (err) {
      this.catchErr(err, this.currentGenre)
    }
    finally {
      this.setLoadingActivities(false)
    }
  }
}

const activitiesStore = new ActivitiesStore();

export { activitiesStore as default };
