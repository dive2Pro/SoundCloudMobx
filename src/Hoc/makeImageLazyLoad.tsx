import * as React from 'react'
import { inject, observer } from 'mobx-react'
import { autorun, IReactionDisposer } from 'mobx'
import { PERFORMANCE_STORE } from '../constants/storeTypes';
import { PerformanceStore } from '../store/PerformanceStore';
import * as ReactDOM from 'react-dom'

const preImage = require('../../public/images/preload.jpg')

interface IMakeImageLazyLoadWrapperProps {
  src: string
  performanceStore: PerformanceStore
  live: boolean
  size: number
  style: { width?: any, height?: any }
}
const asyncLoadImage = (src) => {
  return new Promise((resolve, reject) => {
    const loadImage = new Image()
    loadImage.onload = () => {
      resolve(loadImage.src)
    }
    loadImage.onerror = (err) => {
      reject(err)
    }
    loadImage.src = src
  })
}
function makeImageLazyLoad<Props, State>(
  Comp: React.ComponentClass<Props & { src: string }> | React.StatelessComponent<Props & { src: string }>
): React.ComponentClass<Props> {
  @inject(PERFORMANCE_STORE)
  class MakeImageLazyLoadWrapper extends React.PureComponent<Props & IMakeImageLazyLoadWrapperProps, {}>{
    handlerObserver: IReactionDisposer;
    imgPath: HTMLImageElement;
    loadImage: any;
    retryCount = 3
    state = {
      imageSrc: preImage
    }
    constructor() {
      super()
      this.loadImage = new Image()
    }

    componentDidMount() {
      const imgNode = ReactDOM.findDOMNode(this)

      this.handlerObserver = autorun(() => {
        const { performanceStore: ps, src } = this.props

        if (ps && ps.scrollLimit.length > 0 && src) {
          const [l, h] = ps.scrollLimit
          const y = imgNode.y || 0
          const imgHeight = imgNode.offsetHeight;
          if ((l === h && y <= l) || (y < l && (y + imgHeight > h))) {
            this.caclImg(src)
          }
        } else if (imgNode && !!src === false) {
          // this.img.src = preImage
          this.setState({
            imageSrc: src
          })
        }
      })

    }

    /**
     * 根据给定的size 转换 特定大小的图片地址
     */
    caclImg = (src: string, newSize?: number) => {
      const reg = /(-large)\./g
      let { size, style } = this.props
      size = newSize ? newSize : size
      if (src && (size && size <= 50)) {
        src = src.replace(reg, '-badge\.')

      } else {
        // tslint:disable-next-line:radix
        let width = size, height = size
        let replace

        if (style) {
          width = parseInt(style.width, 10)
          height = parseInt(style.height, 10);
        }

        size = Math.max(Math.ceil(width), Math.ceil(height))
        if (size > 100 && size < 300) {
          replace = 't300x300'
        } else if (size > 300 && size < 500) {
          replace = 'crop'
        } else {
          replace = 't500x500'
        }
        src = src.replace(reg, `-${replace}\.`);
      }

      this.setImagePath(src)
    }


    setImagePath(src) {
      const { live } = this.props
      asyncLoadImage(src)
        .then(path => {

          this.setState({ imageSrc: path })
          if (this.handlerObserver && !live) {
            this.handlerObserver()
          }
        })
        .catch(err => {
          if (this.retryCount-- > 0) {
            this.caclImg(src)
          }
        })

    }

    componentWillReceiveProps(nextProps) {
      if (!!nextProps.src && nextProps.src !== this.state.imageSrc) {
        this.caclImg(nextProps.src, nextProps.size)
      }
    }
    shouldComponentUpdate(nextProps, nextState) {
      if (nextProps.src && this.state.imageSrc !== nextProps.src && nextProps.length === 1) {
        this.caclImg(nextProps.src)
        return false
      }
      return true
    }


    render() {
      return (
        <Comp
          {...this.props}
          src={this.state.imageSrc}
        />
      )
    }
  }
  return MakeImageLazyLoadWrapper
}

export default makeImageLazyLoad