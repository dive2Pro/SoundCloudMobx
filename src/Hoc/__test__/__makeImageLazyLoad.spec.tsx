jest.mock('../../services/Fetch.ts')

import * as React from 'react'
import makeImageLazyLoad from '../makeImageLazyLoad'
import { mount } from 'enzyme'
import ToJson from 'enzyme-to-json'
import { Provider, inject } from 'mobx-react'
import { performanceStore } from '../../store';
// import { when, autorun } from 'mobx'
describe('makeImageLazyLoad HOC', () => {
  it('初始化', () => {
    const tempWrapper = ({ src }) => {
      return <img src={src} alt="some test" />
    }
    const Component = makeImageLazyLoad(tempWrapper);
    const props = {
      src: 'name-large'
    }
    const app = (<Provider performanceStore={performanceStore}><Component {...props} /></Provider>)
    const wrapper = mount(app)
    expect(ToJson(wrapper)).toMatchSnapshot()
    // expect(wrapper.find('img').prop('src')).not.toEqual(props.src)
    expect(wrapper.find('img').prop('src')).toEqual('preload.jpg')
  })

  it('检测src的变化', (done) => {
    const tempWrapper = ({ src }) => {

      return <img src={src} alt="some test" />
    }
    const Component: any = makeImageLazyLoad(tempWrapper);
    const props = {
      src: 'http://img3.imgtn.bdimg.com/it/u=4271053251,2424464488&fm=23&gp=0.jpg',
      performanceStore: performanceStore
    }
    /**
     * @link https://github.com/mobxjs/mobx-react/issues/
     */
    const app = (<Component.wrappedComponent {...props} />)
    const wrapper = mount(app)
    expect(ToJson(wrapper)).toMatchSnapshot()
    performanceStore.setCurrentGenre('Test')
    performanceStore.setScrollLimit(0, 128)
    wrapper.update()

    /**
     * omg
     * @link
        *  http://stackoverflow.com/questions/38308214/react-enzyme-test-componentdidmount-async-call/40875174#40875174
     */
    setImmediate(() => {
      expect(wrapper.update().find(tempWrapper).prop("src")).toEqual(props.src)
      expect(ToJson(wrapper)).toMatchSnapshot()
      // 这里在渲染之前执行
      // console.log(wrapper.instance())
      expect(wrapper.instance().imageSrc).toEqual(props.src)
      //
      const newProps = {
        src: 'http://ooul6pnb3.bkt.clouddn.com/23192045-large.jpg',
        size: 50
      }
      //
      wrapper.setProps(newProps)
      setImmediate(() => {

        let expectedSrc = 'http://ooul6pnb3.bkt.clouddn.com/23192045-badge.jpg'
        expect(ToJson(wrapper)).toMatchSnapshot()
        expect(wrapper.instance().imageSrc).toEqual(expectedSrc)
        expect(wrapper.find(tempWrapper).prop('src')).toEqual(expectedSrc)
        done()
      })

      // expect(wrapper.update().find(tempWrapper).prop("src")).toEqual(newProps.src)

    })

  })
})
