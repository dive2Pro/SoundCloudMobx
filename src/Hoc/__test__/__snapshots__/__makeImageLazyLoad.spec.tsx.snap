exports[`makeImageLazyLoad HOC 初始化 1`] = `
<Provider
  performanceStore={
    PerformanceStore {
      "genre": "Country",
      "getLoadingStateWidthKey": [Function],
      "isError": [Function],
      "isErrorsMap": Object {},
      "isLoadingsByKey": Object {
        "comments": false,
        "post_track_comment": false,
      },
      "scrollLimitByGenre": Object {
        "Country": Array [
          768,
          768,
        ],
        "activities": Array [
          768,
          768,
        ],
      },
    }
  }>
  <inject-MakeImageLazyLoadWrapper-with-performanceStore
    src="name-large">
    <MakeImageLazyLoadWrapper
      performanceStore={
        PerformanceStore {
          "genre": "Country",
          "getLoadingStateWidthKey": [Function],
          "isError": [Function],
          "isErrorsMap": Object {},
          "isLoadingsByKey": Object {
            "comments": false,
            "post_track_comment": false,
          },
          "scrollLimitByGenre": Object {
            "Country": Array [
              768,
              768,
            ],
            "activities": Array [
              768,
              768,
            ],
          },
        }
      }
      src="name-large">
      <tempWrapper
        performanceStore={
          PerformanceStore {
            "genre": "Country",
            "getLoadingStateWidthKey": [Function],
            "isError": [Function],
            "isErrorsMap": Object {},
            "isLoadingsByKey": Object {
              "comments": false,
              "post_track_comment": false,
            },
            "scrollLimitByGenre": Object {
              "Country": Array [
                768,
                768,
              ],
              "activities": Array [
                768,
                768,
              ],
            },
          }
        }
        src="preload.jpg">
        <img
          alt="some test"
          src="preload.jpg" />
      </tempWrapper>
    </MakeImageLazyLoadWrapper>
  </inject-MakeImageLazyLoadWrapper-with-performanceStore>
</Provider>
`;

exports[`makeImageLazyLoad HOC 检测src的变化 1`] = `
<MakeImageLazyLoadWrapper
  performanceStore={
    PerformanceStore {
      "genre": "Country",
      "getLoadingStateWidthKey": [Function],
      "isError": [Function],
      "isErrorsMap": Object {},
      "isLoadingsByKey": Object {
        "comments": false,
        "post_track_comment": false,
      },
      "scrollLimitByGenre": Object {
        "Country": Array [
          768,
          768,
        ],
        "activities": Array [
          768,
          768,
        ],
      },
    }
  }
  src="http://img3.imgtn.bdimg.com/it/u=4271053251,2424464488&fm=23&gp=0.jpg">
  <tempWrapper
    performanceStore={
      PerformanceStore {
        "genre": "Country",
        "getLoadingStateWidthKey": [Function],
        "isError": [Function],
        "isErrorsMap": Object {},
        "isLoadingsByKey": Object {
          "comments": false,
          "post_track_comment": false,
        },
        "scrollLimitByGenre": Object {
          "Country": Array [
            768,
            768,
          ],
          "activities": Array [
            768,
            768,
          ],
        },
      }
    }
    src="preload.jpg">
    <img
      alt="some test"
      src="preload.jpg" />
  </tempWrapper>
</MakeImageLazyLoadWrapper>
`;

exports[`makeImageLazyLoad HOC 检测src的变化 2`] = `
<MakeImageLazyLoadWrapper
  performanceStore={
    PerformanceStore {
      "genre": "Test",
      "getLoadingStateWidthKey": [Function],
      "isError": [Function],
      "isErrorsMap": Object {
        "Country": false,
      },
      "isLoadingsByKey": Object {
        "Country": true,
        "comments": false,
        "post_track_comment": false,
      },
      "scrollLimitByGenre": Object {
        "Country": Array [
          768,
          768,
        ],
        "Test": Array [
          0,
          128,
        ],
        "activities": Array [
          768,
          768,
        ],
      },
    }
  }
  src="http://img3.imgtn.bdimg.com/it/u=4271053251,2424464488&fm=23&gp=0.jpg">
  <tempWrapper
    performanceStore={
      PerformanceStore {
        "genre": "Test",
        "getLoadingStateWidthKey": [Function],
        "isError": [Function],
        "isErrorsMap": Object {
          "Country": false,
        },
        "isLoadingsByKey": Object {
          "Country": true,
          "comments": false,
          "post_track_comment": false,
        },
        "scrollLimitByGenre": Object {
          "Country": Array [
            768,
            768,
          ],
          "Test": Array [
            0,
            128,
          ],
          "activities": Array [
            768,
            768,
          ],
        },
      }
    }
    src="http://img3.imgtn.bdimg.com/it/u=4271053251,2424464488&fm=23&gp=0.jpg">
    <img
      alt="some test"
      src="http://img3.imgtn.bdimg.com/it/u=4271053251,2424464488&fm=23&gp=0.jpg" />
  </tempWrapper>
</MakeImageLazyLoadWrapper>
`;

exports[`makeImageLazyLoad HOC 检测src的变化 3`] = `
<MakeImageLazyLoadWrapper
  performanceStore={
    PerformanceStore {
      "genre": "Test",
      "getLoadingStateWidthKey": [Function],
      "isError": [Function],
      "isErrorsMap": Object {
        "Country": false,
      },
      "isLoadingsByKey": Object {
        "Country": true,
        "comments": false,
        "post_track_comment": false,
      },
      "scrollLimitByGenre": Object {
        "Country": Array [
          768,
          768,
        ],
        "Test": Array [
          0,
          128,
        ],
        "activities": Array [
          768,
          768,
        ],
      },
    }
  }
  size={50}
  src="http://ooul6pnb3.bkt.clouddn.com/23192045-large.jpg">
  <tempWrapper
    performanceStore={
      PerformanceStore {
        "genre": "Test",
        "getLoadingStateWidthKey": [Function],
        "isError": [Function],
        "isErrorsMap": Object {
          "Country": false,
        },
        "isLoadingsByKey": Object {
          "Country": true,
          "comments": false,
          "post_track_comment": false,
        },
        "scrollLimitByGenre": Object {
          "Country": Array [
            768,
            768,
          ],
          "Test": Array [
            0,
            128,
          ],
          "activities": Array [
            768,
            768,
          ],
        },
      }
    }
    size={50}
    src="http://ooul6pnb3.bkt.clouddn.com/23192045-badge.jpg">
    <img
      alt="some test"
      src="http://ooul6pnb3.bkt.clouddn.com/23192045-badge.jpg" />
  </tempWrapper>
</MakeImageLazyLoadWrapper>
`;
